File: 0172_Factorial_Trailing_Zeroes/172.py
def trailingZeroes(n: int) -> int:
    num_of_two = 0
    num_of_five = 0
    for i in range(1, n + 1):
        if i % 2 != 0 and i % 5 != 0:
            continue
        # fac_result = prime_factorization(i)
        # num_of_two += fac_result.count(2)
        # num_of_five += fac_result.count(5)
        res = two_five_cnt(i)
        num_of_two += res[0]
        num_of_five += res[1]
        # print(i, num_of_two, num_of_five)
    return min(num_of_two, num_of_five)

# def prime_factorization(n):
#     divisor = 2
#     factors = []
#     while divisor * divisor <= n:
#         if n % divisor == 0:
#             factors.append(divisor)
#             n //= divisor
#         else:
#             divisor += 1
#     if n > 1:
#         factors.append(n)
#     return factors



def two_five_cnt(n):
    t, f = 0, 0
    while n % 2 == 0:
        n //= 2
        t += 1
    while n % 5 == 0:
        n //= 5 
        f += 1
    return t, f

print(trailingZeroes(3))
print(trailingZeroes(5))
print(trailingZeroes(10))
print(trailingZeroes(30))

File: 0172_Factorial_Trailing_Zeroes/172.cpp
#include <bits/stdc++.h>
using namespace std;
int T = 0, F = 0;
void two_five_cnt(int i)
{
    // vector<int> ans;
    // ans.clear();
    // int t = 0, f = 0;
    while (i % 2 == 0)
    {
        i /= 2;
        T++;
    }
    while (i % 5 == 0)
    {
        i /= 5;
        F++;
    }
    // ans.push_back(t);
    // ans.push_back(f);
    // return ans;
}

int trailingZeroes(int n)
{
    int num_of_two = 0;
    int num_of_five = 0;
    for (int i = 1; i <= n; i++)
    {
        if (i % 2 == 0 || i % 5 == 0)
        {
            // vector<int> res = two_five_cnt(i);
            // cout << i << " " << num_of_two << " " << num_of_five << "\n";
            two_five_cnt(i);
            num_of_two += T;
            num_of_five += F;
            T = 0;
            F = 0;
        }
    }
    return num_of_two < num_of_five ? num_of_two : num_of_five;
}
int main()
{
    cout << trailingZeroes(0) << " " << trailingZeroes(3) << " " << trailingZeroes(5) << " " << trailingZeroes(10) << " " << trailingZeroes(30);
    return 0;
}